devtools/gradle/gradle-application-plugin/src/main/java/io/quarkus/gradle/tasks/QuarkusDev.java

```diff
@@ -66,155 +74,206 @@ public class QuarkusDev extends QuarkusTask {
+    private final SourceSet mainSourceSet;
 
-    protected Configuration quarkusDevConfiguration;
+    private final CompilerOptions compilerOptions = new CompilerOptions();
 
-    private File buildDir;
+    private final Property<File> workingDirectory;
 
-    private String sourceDir;
+    private final Property<Boolean> preventNoVerify;
+    private final Property<Boolean> shouldPropagateJavaCompilerArgs;
+    private final ListProperty<String> args;
+    private final ListProperty<String> jvmArgs;
+    private final ListProperty<String> compilerArgs;
 
-    private String workingDir;
+    private final Set<File> filesIncludedInClasspath = new HashSet<>();
 
-    private List<String> jvmArgs;
+    @SuppressWarnings("unused")
+    @Inject
+    public QuarkusDev(Configuration quarkusDevConfiguration, QuarkusPluginExtension extension) {
+        this("Development mode: enables hot deployment with background compilation", quarkusDevConfiguration, extension);
+    }
 
-    private boolean preventnoverify = false;
+    public QuarkusDev(
+            String name,
+            Configuration quarkusDevConfiguration,
+            @SuppressWarnings("unused") QuarkusPluginExtension extension) {
+        super(name);
+        this.quarkusDevConfiguration = quarkusDevConfiguration;
+        mainSourceSet = getProject().getExtensions().getByType(SourceSetContainer.class)
+                .getByName(SourceSet.MAIN_SOURCE_SET_NAME);
 
-    private List<String> args = new LinkedList<String>();
+        final ObjectFactory objectFactory = getProject().getObjects();
 
-    private List<String> compilerArgs = new LinkedList<>();
+        workingDirectory = objectFactory.property(File.class);
+        workingDirectory.convention(getProject().provider(() -> QuarkusPluginExtension.getLastFile(getCompilationOutput())));
 
-    private CompilerOptions compilerOptions = new CompilerOptions();
+        preventNoVerify = objectFactory.property(Boolean.class);
+        preventNoVerify.convention(false);
 
-    private boolean shouldPropagateJavaCompilerArgs = true;
+        shouldPropagateJavaCompilerArgs = objectFactory.property(Boolean.class);
+        shouldPropagateJavaCompilerArgs.convention(true);
 
-    @Inject
-    public QuarkusDev() {
-        super("Development mode: enables hot deployment with background compilation");
-    }
+        args = objectFactory.listProperty(String.class);
+        compilerArgs = objectFactory.listProperty(String.class);
+        jvmArgs = objectFactory.listProperty(String.class);
 
-    public QuarkusDev(String name) {
-        super(name);
+        // since we have no formal output from this task, changes to the
+        // input always indicates the task is not up-to-date.  in other
+        // words, from the perspective of outputs, the task is never out
+        // of date.  this puts all onus for up-to-date on the inputs
+        getOutputs().upToDateWhen(task -> true);
```

Comment:
I think we never want this task to be up to date as it start the application even if input didn't change

Reply:
Ah yes, you are right.  I forgot about the "specialness" here.  I'll play with that

Reply:
This is an interesting point overall.  The task defines a number of inputs.  You'd really only ever define inputs to either allow input from the user or to facilitate in up-to-date checks.  As far as I can see, most of the inputs defined for this task are not meant for user input and since we do not care about up-to-date I am now wondering why they are inputs at all.  

IMO we should consider which of these inputs are really inputs and unmark the others.

Reply:
Well let me clarify a bit, as there is a distinction between:

1. exposing the thing for user input which just needs the getter/setter
2. the annotations (`@Input`, ...) which are only used for up-to-date checking.

I am asking about both dimensions

Reply:
I would just "try it and see", but I'd prefer to go input-by-input and run the integration tests.  Unfortunately building the plugin and  running the integration tests combined takes roughly 8-9 minutes each time on my machine.  I'll optimistically say 8, so 8 minutes x 9 inputs is not time I want to spend up front ;)